### DISCUSSION TITLE
Installation

### DISCUSSION THREAD

**Discussion by n80894** [Mesabloo/hm-defense#1] [Q&A]
Hello Mesabloo,
I happen to be a fellow young man who absolutely adore the heavy mach defense games. I went through the code to try and clarify for myself what is needed, I only have a small knowledge of coding but I'm still unfamiliar with apk files etc since I've always use apple products. If in any way its possible to play the heavy mach game, I will ensure i do it. I noticed that in the assets readme file (i'll call it ARM) it refers to the general readme file (RM), ARM says that gradlew as well as how to install the files were in RM. Yet in RM all is see it "TODO" leading me to believe you forgot or changed things as you would rather not reveal how to run it in fear of legal trouble? I'm not a hundred percent sure but I want to ask;
Can you help a fellow heavy mach defense enthusiast out and lead me through the steps on how to do it? dm me on instagram @_nnxto or by imessage/ email n8.08@icloud.com. introduce yourself and remind me your trying to help with heavy mach. I would really apreciate it man! or even just some simple instructions on how to work this thing

**Comment by Mesabloo** [comment]
Hello there @n80894!
Thanks for your interest in this project.

However, I have bad news: I initially wanted to remake the whole game, rip the assets for myself, and try and replay the game like this. But the first part -- coding the whole game -- never really happened in the first place. I was able to re-install the original game on an old ipad of mine, and completely forgot the initial goal of this project. Even worse, I managed to get all the assets etc, but some of them are encrypted (e.g. all the files containing all the stats for machines, turrets, levels, etc) which I did not manage to reverse-engineer (in other words: I actually don't know how to get the original data back). This kind of froze my motivation on this project, and...I also have many things to do on a daily basis for now, unfortunately.

I strongly underestimated the amount of work it would be to do all this from scratch. One may say that it's easy when AIs are coded, everything kinda seems to follow (placing turrets isn't hard, spawning machines seems easy enough, displaying the level as well), but I absolutely wanted to be as close as possible to the original game (taking stats into account).
I might be able to get some of them by actually playing the original game (which I still have installed), for example the stats of user machines and turrets, but then I don't know what to put for the remaning part of it (enemy turrets, etc). I already asked someone I knew was into reverse-engineering, but nothing really went on, which is kinda bad for this project...

But the game part of it still isn't easy to do...and I have some other priorities in life as of now.
I am truly sorry for telling you all this. Installing and playing instructions are indeed not written (hence the `TODO`s), because of all of this.

As for the README file in the assets folder, I am indeed worried about any legal implications it may have (even though the original company does not seem to recognize the game anymore as theirs) to redistribute assets without their consent, which I did not manage to have (in fact, they never replied to my email...). At some point I will be trying to put some order in all this mess, but there are...way more assets than I ever thought there would be in this game, which is a bit surprising. I have been able to rip more than 200 image files (all machines, turrets, bases, powers, menus, etc) from an original IPA file. This is so much information that I need to organize at some point into the assets folder. There are still some little problems with that (mainly related to understanding which assets is where in the game -- managed to discover some planned powerups which never made it into the final game), but with time I should be able to resolve them.

As for running the game (well there's nothing to run for now), you seem to be using an iphone (from the "imessage" part of your message -- I may have misunderstood that) and I haven't thought about making it run on iOS. I unfortunately don't have all the required tooling to be able to (which would allow me to create IPA files for example), so I cannot give strong hints on compiling for iOS (I may have also not included iOS as a target in the beginning).
APK files are application files for *Android* only, and cannot be used on iOS (which uses IPA files -- a little bit more complicated to install).


Regards,
Mesabloo

**Reply by DesertedDawn** [reply]
I came here for greatness...... I have been saddened. Good luck to you in your life

**Reply by SpeedyGwen** [reply]
one of my favourite game of my childhoud, I still hope this project will be worked on at some point, meanwhile Imma try to make a recreation of it in scratch (I'll update here if I do it)  the code could be recreated from it and give a push to this project and give more intrest

**Reply by Mesabloo** [reply]
It will be worked on (the code, as I am sorting/organizing all the assets for now). However, this is a process which takes time, and quite a lot actually.
Having to go through all the original assets, understand some of the stats of the original game, understand how to male games in the first place (yes, I do not have much experience in this domain haha). All of this is a bit slow.
I am also trying to document as much as I can on all the steps required to make this repository work (if you check out [here](https://github.com/Mesabloo/hm-defense/tree/main/core/assets/reverse), you will see everything I had to go through to reverse engineering parts of the original game).

Unfortunately, I also have some side projects that I also want to work on.
I strongly apologize for the slow pace this game is (re)made at. This was also one of my favorite childhood games (and I do actually still enjoy it from time to time) and I understand the sad feeling going on.

**Comment by DesertedDawn** [comment]
I wish you luck in your endeavors. If I could provide assistance, I would.
But alas I have no knowledge that would help.

**Comment by Mesabloo** [comment]
Hi there everyone.
This is not really an answer to the original question, but I wanted to give you all a headsup on the status of the project.

I have been working on it for the past two weeks and managed to get quite far (though the hardest part is the game itself, which I haven't started yet).
Here's a quick demo of what I managed to do (new/delete saves also work, though the dialog boxes need to be slightly tweaked): 

https://user-images.githubusercontent.com/22964017/170699765-03174ffb-2e80-4f00-9124-c013456f0c65.mp4

I have updated the main README to include instructions on how to run it on desktop. Mobile support will come later on.
Bugs are to be expected (this is my first real game).

I also decided, after discussing quite a bit with people, that I'll also redistribute the original assets along with the game (there are included in the `core/assets` folder).
Some of them are slightly transformed, and some of them are my own property (i.e. they do not come from the original game itself).

----------------

There is still quite a bit to be worked on, but this will absolutely be done in time (I hope to get a correct prototype that I can show soon).

**Reply by cn912h** [reply]
I am looking forward to this game. I have been looking for it for a long time. I really like it. I will pay attention to you. Come on

**Reply by Mesabloo** [reply]
Unfortunately, this is far from being one of my main projects, and I'm mainly investing time in it when I'm feeling like it (meaning, not that much). Moreover, classes take most of my time currently so I don't have a lot of time (which I use either doing nothing, or working on some of my other projects which I consider more important).
I guess the game will come together eventually, although there are some aspects which are left to be decided/worked on (for example, how to do the animations for the machines).
I am sorry to bring such news (I know it's been quite a while I haven't worked on it), but you'll have to wait for another quite a while.

**Comment by DesertedDawn** [comment]
That's awesome! I'm glad to hear that you've made progress!

**Comment by michaelsvobodadev** [comment]
Hello Mesabloo,

I'm dabbling in game design (sadly more on the theory side, i'm a beginner at C#/Unity3D and game design school dropout really), and thinking of inspiration for some projects i often come back to this series of games i enjoyed way back when. While i'd have really loved to be able to play it again, after poking around while trying to install it i'm actually more interested in the backstage. Do you have any place you share your findings and experiences when working around this project? A discord or other blog type thingy?

My job keeps me a bit too busy right now tbh but i'd really like to go back to game dev in my free time within the year. If I can I'd likely refresh one of my old projects of a top-down action-rpg that would be in some parts inspired by the look and feel of these HM games, specifically HM2 if you played it. Anyways i'd be happy to chat if you're interested. Also applies to anyone else poking around these games tbh i'd just like to find ppl to bounce ideas around with.

Good luck with your projects!

**Reply by Mesabloo** [reply]
Hi Michael.

I do not have such a place where I kind of documented my whole journey through half of this project, but I definitely remember reverse-engineering the game itself to be the most fun part (there should be GDB script files here and there that I used to extract all the original data files from the original game).
Game development really never was my thing. I wanted to try it at some point (and this is how this project somehow came to life) but it seems this isn't really my thing at all. At least I tried, and I have used Kotlin at least.
I used to think that heavy duty game engines such as Unity were basically way overkill for the game, but I also underestimated the amount of animations and stuff in the game (which isn't easy to do actually).
But I'll definitely say that the most time consuming part of the project was to actually retrieve the original files, and understand what's within (which sometimes did not make sense at all).
If you are willing to remake this game in particular, unfortunately I have not found how to interpret some of those data files, so I don't think I can be of any help regarding this.

**Comment by DesertedDawn** [comment]
Here to chime in as well, kinda. I have no actual experience or skills to
help but I'm here for morale support. You got this!


Mainly because I loved this game and I want to play it again lol

**Comment by Lekrod** [comment]
Hi Mesabloo，
I have searching for this game for a very long time（8years?maybe）.Thank you for your effort!I'm glad to know you,you are the real hero.This game is always in my Ipad4(I can't play it as I can't build a new save),just to lissten to the main theme bgm,the bgm in this game is exellent!I really hope that I can play this game again on my computer one day.I can wait !Best wishes for your study!!>w<

**Comment by fdonovan021** [comment]
Is this game playable in its current state? I will also offer moral support for this project to continue!

How exactly do mac users run this game? I read that we need to paste those lines of code in the root of the project... is that in the github download, or in terminal, or in another app. I have the files downloaded but am totally confused how to get this running (never used Github before)

**Reply by Mesabloo** [reply]
Any user you be able to run the game using the commands given in the root README of this repo.
As a first step, you should have cloned this repository somewhere (using the command `git clone https://github.com/mesabloo/hm-defense`, which you can paste in your terminal).
Then, move into the `hm-defense` directory and type (or copy-paste) the commands mentioned in the root README (still in your terminal).
Note that you (may) need a working version of Java on your computer for this to work.
If all goes well, the "game" should be launched in a new window.

As of now, there is mostly nothing to be played. The only things you can do currently is manage saves, open a level and maybe buy upgrades for your base and schedule machines for creation.
Most of the actual gameplay hasn't been implemented... (yet?)
In particular, turrets creation doesn't exist, there are no AIs for your mechs (or your enemy's), there are no upgrades for mechs and no special attacks.
Again...yet?

I don't know if (and when, if it happens) I'll go back to this project. I lost most of my interest in anything, this project being a part of that (and I'm also quite busy most of the day). I didn't think that it would gain that much views, and I'm simply afraid to sadden everybody in this thread by saying that the game may never actually be finished at all.

I guess the hardest part is handling all the animations for mechs/turrets, which I absolutely don't know how to. Another problem is the interpretation of the data files that I managed to extract from the original game, which don't match the actual values when playing the game (as far as I could see).
It would take months to fully reverse-engineer the whole game (not accounting for actually learning how to do those things, which I don't have much interest in to be honest) so as to determine those unknowns regarding all the values. I guess another solution is to make up some stats which seem quite balanced w.r.t. the original game?

**Comment by SpeedyGwen** [comment]
gosh I miss this game, tho I love that this repository still exist, even if its not the most active project I seen over the last few years, I still have a lot of respect for at least trying and keeping it public ♥

**Reply by vanaur** [reply]
So true. A lot of respect for the author too 🖤
